services/offline/src/lib/global-state-service.js

```diff
@@ -0,0 +1,78 @@
+import isEqual from 'lodash/isEqual'
+import PropTypes from 'prop-types'
+import React, { useEffect, useCallback, useContext, useState } from 'react'
+
+// This file creates a redux-like state management service using React context
+// that minimizes unnecessary rerenders that consume the context.
+// See more at https://github.com/amcgee/state-service-poc
+
+const identity = state => state
+
+export const createStore = (initialState = {}) => {
+    const subscriptions = new Set()
+    let state = initialState
+
+    return {
+        getState: () => state,
+        subscribe: callback => {
+            subscriptions.add(callback)
+        },
+        unsubscribe: callback => {
+            subscriptions.delete(callback)
+        },
+        mutate: mutation => {
+            state = mutation(state)
+            for (const callback of subscriptions) {
+                callback(state)
+            }
+        },
+    }
+}
+
+const GlobalStateContext = React.createContext(createStore())
+const useGlobalStateStore = () => useContext(GlobalStateContext)
+
+export const GlobalStateProvider = ({ store, children }) => (
+    <GlobalStateContext.Provider value={store}>
+        {children}
+    </GlobalStateContext.Provider>
+)
+GlobalStateProvider.propTypes = {
+    children: PropTypes.node,
+    store: PropTypes.shape({}),
+}
+
+export const useGlobalState = (selector = identity) => {
+    const store = useGlobalStateStore()
+    const [selectedState, setSelectedState] = useState(
+        selector(store.getState())
+    )
+
+    useEffect(() => {
+        // NEW: deep equality check before updating
+        const callback = state => {
+            const newSelectedState = selector(state)
+            // Second condition handles case where a selected object gets
+            // deleted, but state does not update
+            if (
+                !isEqual(selectedState, newSelectedState) ||
+                selectedState === undefined
+            )
+                setSelectedState(newSelectedState)
```

Comment:
Some components would not rerender properly to reflect updates that an object in state had been deleted, and the second condition handles that.

I think this is what was going on: If `selectedState` was part of an object that gets deleted from the store, `selectedState` becomes `undefined`, and when `callback` gets called, `newSelectedState` is also `undefined` so `setSelectedState` does not get called and any consuming hook does not get updated with new `undefined` state.  Checking if `selectedState === undefined` handles this case, and it does not increase component render counts as verified by render count tests in [`integration.test.js`](https://github.com/dhis2/app-runtime/pull/901/files#diff-b97f31d5db63f4213c857e2461871a7ddbb6d09fbedd9e9af2eb0f4ea2fa1817)

Reply:
Nice edge case catch!